### PR TITLE
Bumps rand to 0.6.X & associated updates

### PR DESCRIPTION
**Pull request by huitseeker**
/cc @apoelstra

### CODE REVIEW

**Comment by real-or-random** [comment]
related: https://github.com/rust-bitcoin/rust-secp256k1/pull/80#issuecomment-437439393 
So, we could stop exporting rand. Not sure if it's a good idea.

**Comment by real-or-random** [comment]
... and depends on the outcome of https://github.com/rust-bitcoin/rust-bitcoin/issues/206

**Comment by apoelstra** [comment]
What would this increase the minimum rustc version to?

**Comment by real-or-random** [comment]
according to their README, "Since version 0.5, Rand requires Rustc version 1.22 or greater." So we can't do this if we stick to 1.20. https://github.com/rust-bitcoin/rust-bitcoin/pull/210

**Comment by real-or-random** [comment]
So as rust-bitcoin requires 1.22 now, there is nothing that prevents us from taking this update.

Do we want to keep exporting rand? If we don't export, this will break downstream code but we'll break that code probably anyway due to update. It's just possible that we'll need to export it again in a few months when rand upgrades requires a newer Rust again...

**Comment by apoelstra** [comment]
What do you mean "we'll need to export it again"? We should keep exporting rand so that users don't have to manually keep track of what version dependencies we're using and change boilerplate code in response to it.

**Comment by apoelstra** [comment]
Also, importantly, if users want to use their own version of `rand` but are required to have access to rand 0.6 types for use with rust-secp, the only simple way to get it is if rust-secp re-exports it. Then they can use `rand` for their stuff and `secp256k1::rand` for our stuff.

**Comment by real-or-random** [comment]
Yes, so we should just keep it.

@huitseeker Thanks for your PR again! So this requires us to bump to to rust 1.22.0. Can you additionally
 - bump 1.14.0 to 1.22.0 in https://github.com/rust-bitcoin/rust-secp256k1/blob/master/.travis.yml#L8,
 - revert 9a4961c8234894bd0ba30895173187636995c184, and
 - bump the minimum Rust version in the README.md?

(or alternatively, make sure that we can make those changes 
https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/ )

**Comment by huitseeker** [comment]
I could have sworn @sgeisler had diagnosed the Windows failure and merged this. Anyway, @real-or-random any thoughts on whether we should include the reversal of  9a4961c?)

**Comment by real-or-random** [comment]
I don't think that the changes to BadRng are necessary: 
 * `next_64` is fine but we could also just stick to `next_u32`. 
 * There are helper  functions provided to implement the other functions: https://docs.rs/rand_core/0.3.1/rand_core/impls/index.html. So we don't need our own implementation of `fill_bytes` and would not rely on `transmute`.
 * Also `try_fill_bytes` can just implemented using `Ok(fill_bytes(...)`, which is also just a single line of code.

I mean in the end this is just a debugging Rng, so there is nothing wrong with changing it but I feel we should not use `transmute` unless we really need to.

For the Windows build: Yes, that's unfortunate, The reason is that the `cc` crate uses a function on Windows, which requires Rust 1.27. (https://doc.rust-lang.org/std/option/enum.Option.html#method.filter). So you're right, we shouldn't revert 9a4961c8234894bd0ba30895173187636995c184 but instead change the comment in the Travis config to explain that `cc` is the reason for disabling Windows builds on 1.22, and note in the README that the restriction to 1.22 is only for the Linux target.

With those two points addresses, this is good to merge for me but I'd like to see an ACK from everybody due to the rust version bump, not only on Linux to 1.22 but in particular on Windows to at least 1.27 (maybe it's in fact even higher, who knows...). @TheBlueMatt? @apoelstra?

**Comment by real-or-random** [comment]
Ok what I'm saying is wrong. We're not bumping on Windows; I guess the build fails already now. But it still will be good to hear your thoughts.

**Comment by huitseeker** [comment]
I'll make a few changes along the lines of your suggestions later today.

**Comment by real-or-random** [comment]
Thanks for the updates! And sorry that I'm so slow at responding here.

Hm, to be honest, now that I see that we need another rand_core then (which is needed by rand anyway though) I'm not convinced anymore that what I suggested is the best approach. But before I ask you to do even more changes and create even more back and forth, I'll suggest waiting for @apoelstra and @TheBlueMatt's opinion on that issue and on the PR in general.

**Comment by huitseeker** [comment]
I'm happy with a 207acef4f59c79289e7ff6a12dff676087db1a50 like style where the relevant `rand_core` functions are pulled within the PR. Just let me know @apoelstra @TheBlueMatt.

File: README.md [hunk 1]
```diff
@@ -20,5 +20,4 @@ Contributions to this library are welcome. A few guidelines:
 * Any breaking changes must have an accompanied entry in CHANGELOG.md
 * No new dependencies, please.
 * No crypto should be implemented in Rust, with the possible exception of hash functions. Cryptographic contributions should be directed upstream to libsecp256k1.
-* This library should always compile with any combination of features on **Rust 1.14**, which is the currently shipping compiler on Debian.
-
+* This library should always compile with any combination of features on **Rust 1.22**, which is the currently shipping compiler on Debian.
```
Comment (apoelstra) [inline README.md]: I think we should drop the reason; Debian actually ships 1.24 or 1.26 or something. Our reasoning for 1.22 is more complicated.

**Comment by apoelstra** [comment]
concept ACK. I think we need an analagous PR for rust-bitcoin which is using rand 0.3 (lol!!) to update to 0.6 or maybe just drop it.

I'm a bit apprehensive about these new `default` functions...or rather, about the `new` functions that they call. I'm unsure whether the library will interact well with instances of these types which are all-bits-zero. I'm OK with this PR but it may be that a future PR removes these things because they can be used to trigger assertation failures or something.

**Comment by huitseeker** [comment]
@apoelstra 
- will drop min-rust-version reason as suggested in https://github.com/rust-bitcoin/rust-secp256k1/pull/92#issuecomment-465583561
- the `new()` functions are already public and untouched by this PR. The `Default` impl materializes the existence of this public no-argument constructor. Happy to tweak.

File: src/ffi.rs [hunk 2]
```diff
@@ -74,6 +74,12 @@ impl PublicKey {
     pub unsafe fn blank() -> PublicKey { mem::uninitialized() }
 }
 
+impl Default for PublicKey{
+    fn default() -> Self{
```
Comment (jonasnick) [inline src/ffi.rs]: nit: misses space between Self and {, same in Secp256k1

File: .travis.yml [hunk 3]
```diff
@@ -16,7 +16,7 @@ matrix:
   #  - we insist on rust 1.14 only for Debian, and
```
Comment (jonasnick) [inline .travis.yml]: Is that comment up to date? doesn't seem like we're using rand 0.4, nor insisting on rust 1.14

**Pull request merged**